jetty-server/src/main/java/org/eclipse/jetty/server/handler/SizeLimitHandler.java

```diff
@@ -0,0 +1,209 @@
+//  All rights reserved. This program and the accompanying materials
+//  are made available under the terms of the Eclipse Public License v1.0
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.server.handler;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpServletResponseWrapper;
+
+import org.eclipse.jetty.http.BadMessageException;
+import org.eclipse.jetty.http.HttpHeader;
+import org.eclipse.jetty.server.HttpInput;
+import org.eclipse.jetty.server.HttpOutput;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+
+/**
+ * <p>A handler that can limit the size of message bodies in requests and responses.</p>
+ * <p>The optional request and response limits are imposed by checking the {@code Content-Length}
+ * header or observing the actual bytes seen by the handler. Handler order is important, in as
+ * much as if this handler is before a the {@link org.eclipse.jetty.server.handler.gzip.GzipHandler},
+ * then it will limit compressed sized, if it as after the {@link org.eclipse.jetty.server.handler.gzip.GzipHandler}
+ * then the limit is applied to uncompressed bytes.
+ * If a size limit is exceeded then {@link BadMessageException} is thrown with a
+ * {@link org.eclipse.jetty.http.HttpStatus#PAYLOAD_TOO_LARGE_413} status.</p>
+ */
+public class SizeLimitHandler extends HandlerWrapper
+{
+    private static final Logger LOG = Log.getLogger(SizeLimitHandler.class);
+
+    private final long _requestLimit;
+    private final long _responseLimit;
+
+    /**
+     * @param requestLimit The request body size limit in bytes or -1 for no limit
+     * @param responseLimit The response body size limit in bytes or -1 for no limit
+     */
+    public SizeLimitHandler(long requestLimit, long responseLimit)
+    {
+        _requestLimit = requestLimit;
+        _responseLimit = responseLimit;
+    }
+
+    private void checkRequestLimit(long size)
+    {
+        if (_requestLimit >= 0 && size > _requestLimit)
+            throw new BadMessageException(413, "Request body is too large: " + size + ">" + _requestLimit);
+    }
+
+    private void checkResponseLimit(long size)
+    {
+        if (_responseLimit >= 0 && size > _responseLimit)
+            throw new BadMessageException(413, "Response body is too large: " + size + ">" + _responseLimit);
```

Comment:
Should these be extensible so the error can be customized. 

Still think 413 is a weird code here because I thought that was only for the request size.